Implement webpushd built-in showNotification and getNotifications

beidson opened the pull request:
#### 9d08f2e924fa8aacf079af47286772d3498f61c0
<pre>
Implement webpushd built-in showNotification and getNotifications
<a href="https://rdar.apple.com/131362923">rdar://131362923</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=276823">https://bugs.webkit.org/show_bug.cgi?id=276823</a>

Reviewed by Andy Estes.

Instead of delegating out notifications-related things to more than one client app implementation,
we should have webpushd just do as much of it all as possible.

A big glaring &quot;FIXME&quot; removed in this patch shows this was long planned, which made this
easier to finally start as some of the pieces were in place.

iOS Web.app-only for now, with some FIXMEs pointing out some missing macOS details.

This is also without tests, for now, as we don&apos;t API test on iOS devices.
The next patch that adds macOS details for notifications will also add API tests.

* Source/WebCore/en.lproj/Localizable.strings:
* Source/WebKit/Configurations/WebKit.xcconfig:

* Source/WebKit/NetworkProcess/Notifications/NetworkNotificationManager.cpp:
(WebKit::NetworkNotificationManager::showNotification):
(WebKit::NetworkNotificationManager::getNotifications):
* Source/WebKit/NetworkProcess/Notifications/NetworkNotificationManager.h:

* Source/WebKit/NetworkProcess/ServiceWorker/WebSWServerConnection.cpp:
(WebKit::WebSWServerConnection::getNotifications):
* Source/WebKit/NetworkProcess/ServiceWorker/WebSWServerConnection.h:
* Source/WebKit/NetworkProcess/ServiceWorker/WebSWServerConnection.messages.in:

* Source/WebKit/Platform/spi/Cocoa/UserNotificationsSPI.h: Added.
* Source/WebKit/Platform/spi/ios/UIKitSPI.h:

* Source/WebKit/Resources/SandboxProfiles/ios/com.apple.WebKit.webpushd.sb.in:
* Source/WebKit/Scripts/process-entitlements.sh:

* Source/WebKit/UIProcess/WebPreferences.cpp:
(WebKit::WebPreferences::updateBoolValueForKey):

* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

* Source/WebKit/WebProcess/Storage/WebSWClientConnection.cpp:
(WebKit::WebSWClientConnection::getNotifications):

* Source/WebKit/webpushd/PushClientConnection.h:
* Source/WebKit/webpushd/PushClientConnection.messages.in:
* Source/WebKit/webpushd/PushClientConnection.mm:
(WebPushD::PushClientConnection::showNotification):
(WebPushD::PushClientConnection::getNotifications):
(WebPushD::PushClientConnection::associatedWebClipTitle const):

* Source/WebKit/webpushd/WebPushDaemon.h:
* Source/WebKit/webpushd/WebPushDaemon.mm:
(WebPushD::platformShouldPlaySound):
(WebPushD::platformDefaultActionBundleIdentifier):
(WebPushD::platformNotificationCenterBundleIdentifier):
(WebPushD::platformNotificationSourceForDisplay):
(WebPushD::WebPushDaemon::showNotification):
(WebPushD::WebPushDaemon::getNotifications):

Canonical link: <a href="https://commits.webkit.org/281190@main">https://commits.webkit.org/281190@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/295b20811918173bf1f3380f17f269a362726588

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/59053 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/38381 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/11548 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/62773 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/9496 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/61182 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/46032 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/9702 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/47834 "") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/6781 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/61083 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/35905 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/51103 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/28692 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/58579 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/32627 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/8373 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/8500 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/54582 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/8653 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/64471 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/2965 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/8609 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/55155 "") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/2975 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/51110 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/55260 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/13070 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/2491 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/34210 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/35294 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/36379 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/35040 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

emw-apple commented on Source/WebKit/Configurations/WebKit.xcconfig:
```diff
@@ -243,8 +243,13 @@ WK_WRITING_TOOLS_LDFLAGS_maccatalyst_MACOS_SINCE_1500 = -weak_framework WritingT
 WK_WRITING_TOOLS_LDFLAGS_MACOS_SINCE_1500 = -weak_framework WritingTools;
 
+WK_USERNOTIFICATIONS_LDFLAGS = $(WK_USERNOTIFICATIONS_LDFLAGS_$(WK_PLATFORM_NAME));
+WK_USERNOTIFICATIONS_LDFLAGS_ = ;
```
I think you can remove this line. For platform names other than the three you list below, the build setting will be implicitly empty by default.

aestes commented on Source/WebKit/Platform/spi/Cocoa/UserNotificationsSPI.h:
```diff
@@ -0,0 +1,52 @@
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
```
It's atypical to put `#pragma once` in an Objective-C header (since they should always be `#import`ed)

beidson committed:
Implement webpushd built-in showNotification and getNotifications
rdar://131362923
https://bugs.webkit.org/show_bug.cgi?id=276823

Reviewed by Andy Estes.

Instead of delegating out notifications-related things to more than one client app implementation,
we should have webpushd just do as much of it all as possible.

A big glaring "FIXME" removed in this patch shows this was long planned, which made this
easier to finally start as some of the pieces were in place.

iOS Web.app-only for now, with some FIXMEs pointing out some missing macOS details.

This is also without tests, for now, as we don't API test on iOS devices.
The next patch that adds macOS details for notifications will also add API tests.

* Source/WebCore/en.lproj/Localizable.strings:
* Source/WebKit/Configurations/WebKit.xcconfig:

* Source/WebKit/NetworkProcess/Notifications/NetworkNotificationManager.cpp:
(WebKit::NetworkNotificationManager::showNotification):
(WebKit::NetworkNotificationManager::getNotifications):
* Source/WebKit/NetworkProcess/Notifications/NetworkNotificationManager.h:

* Source/WebKit/NetworkProcess/ServiceWorker/WebSWServerConnection.cpp:
(WebKit::WebSWServerConnection::getNotifications):
* Source/WebKit/NetworkProcess/ServiceWorker/WebSWServerConnection.h:
* Source/WebKit/NetworkProcess/ServiceWorker/WebSWServerConnection.messages.in:

* Source/WebKit/Platform/spi/Cocoa/UserNotificationsSPI.h: Added.
* Source/WebKit/Platform/spi/ios/UIKitSPI.h:

* Source/WebKit/Resources/SandboxProfiles/ios/com.apple.WebKit.webpushd.sb.in:
* Source/WebKit/Scripts/process-entitlements.sh:

* Source/WebKit/UIProcess/WebPreferences.cpp:
(WebKit::WebPreferences::updateBoolValueForKey):

* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

* Source/WebKit/WebProcess/Storage/WebSWClientConnection.cpp:
(WebKit::WebSWClientConnection::getNotifications):

* Source/WebKit/webpushd/PushClientConnection.h:
* Source/WebKit/webpushd/PushClientConnection.messages.in:
* Source/WebKit/webpushd/PushClientConnection.mm:
(WebPushD::PushClientConnection::showNotification):
(WebPushD::PushClientConnection::getNotifications):
(WebPushD::PushClientConnection::associatedWebClipTitle const):

* Source/WebKit/webpushd/WebPushDaemon.h:
* Source/WebKit/webpushd/WebPushDaemon.mm:
(WebPushD::platformShouldPlaySound):
(WebPushD::platformDefaultActionBundleIdentifier):
(WebPushD::platformNotificationCenterBundleIdentifier):
(WebPushD::platformNotificationSourceForDisplay):
(WebPushD::WebPushDaemon::showNotification):
(WebPushD::WebPushDaemon::getNotifications):

Canonical link: https://commits.webkit.org/281190@main
